api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java

```diff
@@ -330,5 +340,212 @@ public void purgeDiagnosis_shouldPurgeDiagnosis() {
+		assertEquals(1, diagnosisAttribute.getId());
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnosisAttributeByUuid(String)
+	 */
+	@Test
+	public void getDiagnosisAttributeByUuid_shouldReturnNullIfNoDiagnosisAttributeHasTheProvidedUuid() {
+		assertNull(diagnosisService.getDiagnosisAttributeByUuid("001612a5-9f2a-4a05-9384-755679f75647"));
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	@Test
+	public void saveDiagnosis_shouldSaveTheDiagnosisWithTheProvidedAttributes() {
+		final String NAMESPACE = "namespace", FORMFIELD_PATH = "formFieldPath";
+		final DiagnosisAttributeType diagnosisAttributeType = Context.getDiagnosisService()
+				.getDiagnosisAttributeTypeByUuid("949daf5b-a83e-4b65-b914-502a553243d3");
+		DiagnosisAttribute diagnosisAttribute = new DiagnosisAttribute();
+		diagnosisAttribute.setId(2);
+		diagnosisAttribute.setUuid("3cc9ad27-a4e1-4a08-acd5-fb6b847d71d0");
+		diagnosisAttribute.setAttributeType(diagnosisAttributeType);
+		diagnosisAttribute.setCreator(diagnosisAttributeType.getCreator());
+		diagnosisAttribute.setVoided(false);
+		diagnosisAttribute.setValueReferenceInternal("Diagnosis Attribute Reference");
+		String uuid = "3c6422d8-7bdd-4e20-bd1b-a590f084db08";
+		Condition condition = conditionService.getConditionByUuid("2cc6880e-2c46-15e4-9038-a6c5e4d22fb7");
+		Encounter encounter = encounterService.getEncounterByUuid("y403fafb-e5e4-42d0-9d11-4f52e89d123r");
+		Patient patient = patientService.getPatient(2);
+		Diagnosis diagnosis = new Diagnosis();
+		diagnosis.setUuid(uuid);
+		diagnosis.setEncounter(encounter);
+		diagnosis.setCondition(condition);
+		diagnosis.setCertainty(ConditionVerificationStatus.CONFIRMED);
+		diagnosis.setPatient(patient);
+		diagnosis.setRank(1);
+		diagnosis.setVoided(false);
+		diagnosis.setFormField(NAMESPACE, FORMFIELD_PATH);
+		diagnosis.setAttribute(diagnosisAttribute);
+		diagnosisService.save(diagnosis);
+		assertNotNull(diagnosis.getId(), "Successfully Saved Diagnosis");
+		Diagnosis savedDiagnosis = diagnosisService.getDiagnosisByUuid(uuid);
+		assertEquals(uuid, savedDiagnosis.getUuid());
+		assertEquals(condition, savedDiagnosis.getCondition());
+		assertEquals(encounter, savedDiagnosis.getEncounter());
+		assertEquals(patient, savedDiagnosis.getPatient());
+		assertEquals(ConditionVerificationStatus.CONFIRMED, savedDiagnosis.getCertainty());
+		assertEquals(1, savedDiagnosis.getRank());
+		assertEquals(false, savedDiagnosis.getVoided());
+		assertEquals(NAMESPACE + "^" + FORMFIELD_PATH, savedDiagnosis.getFormNamespaceAndPath());
+		assertThat(savedDiagnosis.getAttributes(), hasItem(diagnosisAttribute));
+		assertThat(diagnosisAttribute.getAttributeType().getName(), is("Differential Diagnosis"));
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#purgeDiagnosis(Diagnosis)
+	 */
+	@Test
+	public void purgeDiagnosis_shouldPurgeTheDiagnosisAndTheAssociatedAttributes() {
+		final String uuid = "68802cce-6880-17e4-6880-a68804d22fb7";
```

Comment:
Did you take a look at our naming convention for constants? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

Reply:
I also did not mean purging the diagnosis, but rather removing attributes from a diagnosis.

Reply:
@dkayiwa I thought the attributes are detached on the deletion of the diagnosis. Would retiring a diagnosis have the same effect?

Reply:
Just like one can add attributes to a diagnosis, they should also be able to remove them.

Reply:
Alright, thanks.